Ignore endpoints for ingresses that don't have status.LoadBalancer.Ingress defined

jadolg opened the pull request:
<!--
    Please read https://github.com/kubernetes-sigs/external-dns#contributing before submitting
    your pull request. Please fill in each section below to help us better prioritize your pull request. Thanks!
-->

**Description**

<!-- Please provide a summary of the change here. -->
If there's a hostname defined but there's no target to point it to, external-dns should skip it instead of deleting it.
We have had a failing ingress controller that removed the status from the ingress object when it got OOMKilled. This resulted in external-dns deleting DNS entries even though the load balancer address was not changed.
<!-- Please link to all GitHub issue that this pull request implements(i.e. Fixes #123) -->
Fixes #2677

**Checklist**

- [ ] Unit tests updated
- [ ] End user documentation updated

hvoigt commented on plan/plan.go:
```diff
@@ -146,6 +146,10 @@ func (p *Plan) Calculate() *Plan {
 
 	for _, topRow := range t.rows {
 		for _, row := range topRow {
+			if endpoint.EndpointsHaveEmptyTargets(row.candidates) {
```
How about adding a comment here
```suggestion
                        // We ignore defined Endpoints with unknown targets, as there seems to be
                        // intention for the DNS entry, but it is not known where to point them.  
			if endpoint.EndpointsHaveEmptyTargets(row.candidates) {
```

hvoigt commented on source/ingress.go:
```diff
@@ -129,6 +129,10 @@ func (sc *ingressSource) Endpoints(ctx context.Context) ([]*endpoint.Endpoint, e
 
 		ingEndpoints := endpointsFromIngress(ing, sc.ignoreHostnameAnnotation, sc.ignoreIngressTLSSpec, sc.ignoreIngressRulesSpec)
 
+		if sc.fqdnTemplate != nil && endpoint.EndpointsHaveEmptyTargets(ingEndpoints) {
```
Another comment maybe:
```suggestion
                // Template configurations are a fallback to determine endpoint targets 
		if sc.fqdnTemplate != nil && endpoint.EndpointsHaveEmptyTargets(ingEndpoints) {
```

hvoigt reviewed (approved):
LGTM, just a few comments.

AmitKumarDas commented on source/ingress.go:
```diff
@@ -226,6 +231,24 @@ func (sc *ingressSource) setDualstackLabel(ingress *networkv1.Ingress, endpoints
 	}
 }
 
+func endpointsForIngressHostame(hostname string, targets endpoint.Targets, ttl endpoint.TTL, providerSpecific endpoint.ProviderSpecific, setIdentifier string) []*endpoint.Endpoint {
```
```suggestion
func endpointsForIngressHostname(hostname string, targets
endpoint.Targets, ttl endpoint.TTL, providerSpecific endpoint.ProviderSpecific, setIdentifier string)
```

AmitKumarDas replied:
```suggestion
func endpointsForIngressHostname(hostname string, targets endpoint.Targets, ttl endpoint.TTL, providerSpecific endpoint.ProviderSpecific, setIdentifier string) []*endpoint.Endpoint {
```

AmitKumarDas commented on plan/plan.go:
```diff
@@ -146,6 +146,12 @@ func (p *Plan) Calculate() *Plan {
 
 	for _, topRow := range t.rows {
 		for _, row := range topRow {
+			// We ignore defined Endpoints with unknown targets, as there seems to be
```
suggest: Can we provide in code comments a real world use case that explains this scenario? Otherwise this new condition might change the way existing deployments work (probably leading to issues, incidents, etc.)

We might also be prompted to make this a feature flag if this is not a bug.

jadolg committed:
Ignore endpoints for ingresses that don't have status.LoadBalancer.Ingress defined

If there's a hostname defined but there's no target to point it to, external-dns should skip it instead of deleting it.
We have had a failing ingress controller that removed the status from the ingress object when it got OOMKilled. This resulted in external-dns deleting DNS entries even though the load balancer address was not changed.
